reindexer/reindex_worker/src/main/scala/uk/ac/wellcome/platform/reindex_worker/services/ReindexService.scala

```diff
@@ -31,130 +30,61 @@ class ReindexService @Inject()(dynamoDBClient: AmazonDynamoDB,
+    // not to be a problem.
+    val queryResults: List[Either[DynamoReadError, HybridRecord]] =
+      Scanamo.exec(dynamoDBClient)(
+        index.query(
+          'reindexShard -> reindexJob.shardId and
+            KeyIs('reindexVersion, LT, reindexJob.desiredVersion)
         )
-      case Right(hybridRecord: HybridRecord) => {
-        getRecord(hybridRecord).flatMap { record =>
-          updateRecord(record, desiredVersion)
-        }
-      }
-    }
-
-    Future
-      .sequence(updates)
-      .map(_ => {
-        info(s"ReindexTargetService completed batch of ${updates.length}")
+      )
 
-        metricsSender.incrementCount(
-          "reindex-updated-items",
-          updates.length
+    val outdatedRecords = queryResults.map {
+      case Left(err: DynamoReadError) => {
+        warn(s"Failed to read Dynamo records for $reindexJob: $err")
+        throw GracefulFailureException(
+          new RuntimeException(s"Error in the DynamoDB query: $err")
         )
-
-        true
-      })
-      .recover {
-        case _ => false
       }
-  }
-
-  private def createScanamoQueryRequest(
-    reindexJob: ReindexJob): ScanamoQueryRequest =
-    ScanamoQueryRequest(
-      dynamoConfig.table,
-      Some(gsiName),
-      Query(
-        AndQueryCondition(
-          KeyEquals('reindexShard, reindexJob.shardId),
-          KeyIs('reindexVersion, LT, reindexJob.desiredVersion)
-        )),
-      ScanamoQueryOptions.default
-    )
-
-  def runReindex(reindexJob: ReindexJob)(
-    implicit evidence: DynamoFormat[HybridRecord]): Future[Unit] = {
-
-    info(s"ReindexTargetService running $reindexJob")
-
-    val scanamoQueryRequest = createScanamoQueryRequest(reindexJob)
+      case Right(r: HybridRecord) => r
+    }
 
-    val ops: ScanamoOps[List[Future[Boolean]]] = scanamoQueryStreamFunction(
-      queryRequest = scanamoQueryRequest,
-      resultFunction = updateQueryResults(reindexJob.desiredVersion)
-    )
+    // Then we PUT all the records.  It might be more efficient to do a
+    // bulk update, but this will do for now.
+    val updates: List[Future[Unit]] = outdatedRecords.map { hybridRecord =>
+      val updatedRecord =
+        hybridRecord.copy(reindexVersion = reindexJob.desiredVersion)
+      versionedDao.updateRecord[HybridRecord](updatedRecord)(evidence =
+                                                               evidence,
+                                                             versionUpdater =
+                                                               versionUpdater)
+    }
 
-    Future
-      .sequence(Scanamo.exec(dynamoDBClient)(ops))
-      .map(result =>
-        if (result.contains(false)) {
-          throw GracefulFailureException(
-            new RuntimeException(
-              "Not all records were successfully processed!"
-            ))
-        } else {
-          info(s"Successfully processed reindex job $reindexJob")
-      })
+    Future.sequence(updates)
```

Comment:
This returns `Future[List[Unit]]`, which may not behave correctly with the for comprehension in the worker service.

Reply:
Also, I don't log anything here. 🙈